src/mpi/coll/src/coll_impl.c

```diff
@@ -259,10 +259,10 @@ void MPIR_Coll_host_buffer_swap_back(void *host_sendbuf, void *host_recvbuf, voi
     request->u.nbc.coll.host_recvbuf = host_recvbuf;
     if (host_recvbuf) {
         request->u.nbc.coll.user_recvbuf = in_recvbuf;
-        request->u.nbc.coll.count = count;
-        request->u.nbc.coll.datatype = datatype;
-        MPIR_Datatype_add_ref_if_not_builtin(datatype);
     }
+    request->u.nbc.coll.count = count;
+    request->u.nbc.coll.datatype = datatype;
+    MPIR_Datatype_add_ref_if_not_builtin(datatype);
```

Comment:
I trust this is correct, but where in the send path is the datatype used again? Not clear from these changes.

Reply:
In `MPIR_gpu_host_free` of `send_buf`, it need query the datatype to find `true_lb` and `extent`, in order to find the actual buffer address to free.

Reply:
OK, got it.